libvast/src/arrow_table_slice.cpp

```diff
@@ -585,387 +603,196 @@ class index_applier {
 
-  using size_type = typename super::size_type;
-
-  using value_type = typename super::value_type;
-
-  using array_ptr = std::shared_ptr<arrow::Array>;
-
-  experimental_container_view(type element_type, array_ptr arr, int32_t offset,
-                              int32_t length)
-    : element_type_(std::move(element_type)),
-      offset_(offset),
-      length_(length),
-      arr_{std::move(arr)} {
-    // nop
-  }
+} // namespace legacy
 
-  value_type at(size_type row) const override {
-    auto adjusted_row = row + detail::narrow_cast<size_type>(offset_);
-    if constexpr (std::is_same_v<value_type, data_view>) {
-      return value_at(element_type_, *arr_, adjusted_row);
-    } else {
-      using expected_type = std::pair<data_view, data_view>;
-      static_assert(std::is_same_v<value_type, expected_type>);
-      if (const auto* dt = caf::get_if<record_type>(&element_type_)) {
-        if (dt->num_fields() == 2) {
-          const auto& arr = static_cast<const arrow::StructArray&>(*arr_);
-          auto key_arr = arr.field(0);
-          auto value_arr = arr.field(1);
+data_view value_at(const type& type, const std::same_as<arrow::Array> auto& arr,
+                   int64_t row) noexcept;
+
+template <concrete_type Type>
+view<type_to_data_t<Type>>
+value_at([[maybe_unused]] const Type& type,
+         const type_to_arrow_array_t<Type>& arr, int64_t row) noexcept {
+  VAST_ASSERT(!arr.IsNull(row));
+  if constexpr (detail::is_any_v<Type, bool_type, count_type, real_type>) {
+    return arr.GetView(row);
+  } else if constexpr (std::is_same_v<Type, integer_type>) {
+    return integer{arr.GetView(row)};
+  } else if constexpr (std::is_same_v<Type, duration_type>) {
+    VAST_ASSERT(
+      caf::get<type_to_arrow_type_t<duration_type>>(*arr.type()).unit()
+      == arrow::TimeUnit::NANO);
+    return duration{arr.GetView(row)};
+  } else if constexpr (std::is_same_v<Type, time_type>) {
+    VAST_ASSERT(caf::get<type_to_arrow_type_t<time_type>>(*arr.type()).unit()
+                == arrow::TimeUnit::NANO);
+    return time{} + duration{arr.GetView(row)};
+  } else if constexpr (std::is_same_v<Type, string_type>) {
+    const auto str = arr.GetView(row);
+    return {str.data(), str.size()};
+  } else if constexpr (std::is_same_v<Type, pattern_type>) {
+    const auto& storage
+      = caf::get<type_to_arrow_array_t<string_type>>(*arr.storage());
+    return view<type_to_data_t<pattern_type>>{
+      value_at(string_type{}, storage, row)};
+  } else if constexpr (std::is_same_v<Type, address_type>) {
+    const auto& storage
+      = static_cast<const arrow::FixedSizeBinaryArray&>(*arr.storage());
+    VAST_ASSERT(storage.byte_width() == 16);
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    const auto* bytes = storage.raw_values() + (row * 16);
+    return address::v6(std::span<const uint8_t, 16>{bytes, 16});
+  } else if constexpr (std::is_same_v<Type, subnet_type>) {
+    const auto& storage
```

Comment:
It's technically unnecessary to do this workaround every time. We could consider adding a `type_to_arrow_array_storage_t<T>` or so. At the very least there should be a TODO comment for that here.

Reply:
Would that not work before the if-else chain?
```
using array_type = type_to_arrow_array_t<Type>;
const auto& storage = static_cast<const array_type&>(*arr.storage());
```

Reply:
No, the storage array is of a different array type, because it matches the extension type's underlying storage type and not the extension type itself. So for a `type_to_arrow_array_t<subnet_type>`, the matching storage array type is `arrow::StructArray`.